### PR TITLE
Closes issue #29 include command can raise unhandled exception

### PR DESCRIPTION
**Pull request by Rudedog9d**
Adds error handling for malformed files being read by the include command. Also adds support for multi-line commands in the included files and fixes a false "recursive include" bug.

### CODE REVIEW

File: pypsi/commands/include.py [hunk 1]
```diff
@@ -65,28 +67,53 @@ def run(self, shell, args):
...
         except (OSError, IOError) as e:
             self.error(shell, "error opening file ", path, ": ", str(e), '\n')
+            self.stack.pop()
             return -1
 
-        for line in fp:
-            shell.execute(line.strip())
-            ifile.line += 1
+        try:
+            self.lines = [line for line in fp]
```
Comment (ameily) [inline pypsi/commands/include.py]: This is load the entire file into memory, which isn't ideal. It'd be nice to read the file line by line.

File: pypsi/commands/include.py [hunk 2]
```diff
@@ -65,28 +67,53 @@ def run(self, shell, args):
 
     def include_file(self, shell, path):
         fp = None
-        ifile = IncludeFile(path)
+        self.ifile = IncludeFile(path)
```
Comment (ameily) [inline pypsi/commands/include.py]: What's the point of making `ifile` an instance attribute?

**Comment by ameily** [comment]
I think the actual include functionality needs to be implemented in the shell, Moving the functionality into the shell 

 * Mirror the `cmdloop` function.
 * Redirect sys.stdin to the file object returned by opening the include file.
 * As you walk through the file you can catch UnicodeError and then close the file
 * Reaching the end of the file will raise a EOFError
 * The include command should be pretty simple: track the open file stack and feed each file into the shell.

File: pypsi/commands/include.py (hunk 2 above)
Comment (Rudedog9d) [inline pypsi/commands/include.py]: `ifile` keeps track of the line number, I made it an instance var so I could track what line I was on in the `get_next_line` function.

**Comment by ameily** [comment]
It's generally bad practice to use exceptions to implement functionality.

 * Use `instanceof()` or `hasattr()` instead of catching `AttributeError`
 * Use `index < len(x)` instead of catching `IndexError`

File: pypsi/commands/include.py [hunk 3]
```diff
@@ -27,6 +27,7 @@ def __init__(self, path, line=1):
         self.name = os.path.basename(path)
         self.abspath = os.path.abspath(path)
         self.line = line
+        self.ifile = None
```
Comment (ameily) [inline pypsi/commands/include.py]: I think you can get rid of this now.

File: pypsi/shell.py [hunk 4]
```diff
@@ -194,6 +194,63 @@ def on_tokenize(self, tokens, origin):
...
+        rc = 1
+        eof = False
+
+        # set STDIN to the file
+        stdin = sys.stdin._get_target()
+        sys.stdin._proxy(file)
+
+        try:
+            while self.running and not eof:
+                try:
```
Comment (ameily) [inline pypsi/shell.py]: I think you can wrap the input and execution in a single try...except to catch KeyboardInterrupt, EOFError, and SystemExit only once.

**Comment by Sidneys1** [comment]
A bug was introduced where `shell.running` is not instantiated in `shell.__init__`, causing `'Shell' object has no attribute 'running'` if the `include` command is used before `cmd_loop` is called.

**Pull request merged**